designs/v1alpha6.md

```diff
@@ -0,0 +1,99 @@
+
+### Implementation Plan
+
+The Karpenter codebase currently references v1alpha5.Provisioner in most of its controllers. While v1alpha5 and v1alpha6 are not roundtrippable, we can restore the broken relationship between v1alpha5.Provisioner.spec.provider and v1alpha6.AWSNodeTemplate in memory. This can be achieved by json encoding provisioner.spec.provider into an annotation karpenter.sh/compatibility/provider. Any area of code that needs to maintain compatibility will be required to parse out these values. Luckily, this is contained to the AWS cloud provider, as spec.provider predates prototyping efforts from other vendors.
+
+Alternatively, we can choose to replicate existing v1alpha5 controllers into a new folder controllers/v1alpha6/... where each set of versioned controllers operates independently on Provisioners of different API version. This will result in a much larger volume of code in the short term, but keep the complexity isolated. Developers will need to be aware of this duplication as long as it exists, and backport fixes to both places where necessary. Because of this, we will limit the number of Karpenter releases that support both versions and eventually deprecate support for v1alpha5.
+
+Either approach impacts the velocity of the project — we will need to decide when is the best time to make these changes.
+
+## Proposed Changes
+
+### Rename: Provisioner → NodeProvisioner
+
+This change avoids the roundtrippable problem mentioned above. Further, Provisioner is an overloaded term, colliding with storage provisioning, control plane provisioning, network provisioning. The name NodeProvisioner has nice alignment with NodeTemplate, and potentially in the future NodeGroup. NodeProvisioner and NodeGroup would each be responsible for managing a set of nodes, but with differing rules for when capacity is launched. They can be thought of as dynamic and static capacity.
+
+Naming confusion:
+
+* https://kubernetes.slack.com/archives/C02SFFZSA2K/p1669605517500229?thread_ts=1669289308.636989&cid=C02SFFZSA2K
+
+### Rename: Move provisioner.spec to provisioner.spec.template
+
+Currently fields that control node properties, such as Labels, Taints, StartupTaints, Requirements, Kubelet, ProviderRef, are top level members of provisioner.spec. These are siblings of fields that control provisioning behavior, such as Limits, TTLSecondsUntilExpired, and Weight.
+
+Instead, we should separate out properties of the node in the same way that pod properties are separated from their parent controllers. This has a nice effect of simultaneously creating a place for users to define node metadata. This design mirrors that of kubernetes objects that contain podtemplatespecs, like Deployment, StatefulSet, Job, and CronJob.
+```
+spec:
+  ttlSecondsUntilExpired: ...
```

Comment:
**Opinion**

Consider moving this into a containing field such as `nodeLifecycle`. We could do that for `v1beta1` though as the API change would survive round trips.